transport/transport-runtime/src/main/java/com/google/android/datatransport/runtime/scheduling/jobscheduling/WorkManagerScheduler.java

```diff
@@ -0,0 +1,114 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.android.datatransport.runtime.scheduling.jobscheduling;
+
+import static android.util.Base64.DEFAULT;
+import static android.util.Base64.encodeToString;
+
+import android.content.Context;
+import android.os.Build;
+import androidx.annotation.RequiresApi;
+import androidx.work.Data;
+import androidx.work.OneTimeWorkRequest;
+import androidx.work.WorkInfo;
+import androidx.work.WorkManager;
+import androidx.work.WorkRequest;
+import com.google.android.datatransport.runtime.TransportContext;
+import com.google.android.datatransport.runtime.logging.Logging;
+import com.google.android.datatransport.runtime.scheduling.persistence.EventStore;
+import com.google.android.datatransport.runtime.util.PriorityMapping;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+@RequiresApi(api = Build.VERSION_CODES.UPSIDE_DOWN_CAKE)
+public class WorkManagerScheduler implements WorkScheduler {
+  private static final String LOG_TAG = "WorkManagerScheduler";
+
+  static final String ATTEMPT_NUMBER = "attemptNumber";
+  static final String BACKEND_NAME = "backendName";
+  static final String EVENT_PRIORITY = "priority";
+  static final String EXTRAS = "extras";
+  private final Context context;
+
+  private final EventStore eventStore;
+
+  private final SchedulerConfig config;
+
+  public WorkManagerScheduler(
+      Context applicationContext, EventStore eventStore, SchedulerConfig config) {
+    this.context = applicationContext;
+    this.eventStore = eventStore;
+    this.config = config;
+  }
+
+  @Override
+  public void schedule(TransportContext transportContext, int attemptNumber) {
+    schedule(transportContext, attemptNumber, false);
+  }
+
+  @Override
+  public void schedule(TransportContext transportContext, int attemptNumber, boolean force) {
+    WorkManager manager = WorkManager.getInstance(context);
+
+    int jobId = WorkScheduler.getJobId(context, transportContext);
+    if (!force) {
+      try {
+        for (WorkInfo info : manager.getWorkInfosByTag("transport-" + jobId).get()) {
+          if (!info.getState().isFinished()) {
+            Logging.d(
+                LOG_TAG,
+                "Upload for context %s is already scheduled. Returning...",
+                transportContext);
+            return;
+          }
+        }
+      } catch (Exception e) {
```

Comment:
Could you add a comment here on why is OK to ignore any exception?

Also, is there a more specific Exception we could use ?

Reply:
The WorkManager API shouldn't be throwing the handful of exceptions on the .get(), but if something changes, I've made it rethrow a runtime exception.

Reply:
It has 3 possible exception (from the refdocs)

> Throws:
> CancellationException – if the computation was cancelled
> ExecutionException – if the computation threw an exception
> InterruptedException – if the current thread was interrupted while waiting

Are we OK crashing the app if any of these happen?